python/cudf/cudf/core/index.py

```diff
@@ -2703,6 +2706,144 @@ def categories(self):
+        # the below adjusts for this
+        if cupy.all(cupy.mod(bin_edges, 1) == 0):
+            bin_edges = bin_edges.astype(int)
+        left_col = bin_edges[:-1]
+        right_col = bin_edges[1:]
+    else:
+        if freq and periods and end:
+            start = end - (freq * periods)
+        if freq and periods and start:
+            end = freq * periods + start
+        left_col = cupy.arange(start, end, freq)
+        if end is not None:
+            end = end + 1
+        if start is not None and freq is not None:
+            new_freq = start + freq
+        elif start is not None and freq is None:
+            new_freq = start
+        right_col = cupy.arange(new_freq, end, freq)
+        if len(left_col) != len(right_col):
+            if end and freq is not None:
+                end = end - freq
+            left_col = cupy.arange(start, end, freq)
+        if len(right_col) == 0 or len(left_col) == 0:
+            return pd.IntervalIndex([], closed=closed)
+
+    interval_col = column.build_interval_column(
+        left_col, right_col, closed=closed
+    )
+    return IntervalIndex(interval_col)
+
+
+class IntervalIndex(GenericIndex):
+    """
+    Immutable index of intervals that are closed on the same side.
+
+    Parameters
+    data : array-like (1-dimensional)
+    Array-like containing Interval objects from which to build the
+    IntervalIndex.
+
+    closed : {‘left’, ‘right’, ‘both’, ‘neither’}, default ‘right’
+    Whether the intervals are closed on the left-side, right-side,
+    both or neither.
+
+    dtype : dtype or None, default None
+    If None, dtype will be inferred.
+
+    copy : bool, default False
+    Copy the input data.
+
+    name : object, optional
+    Name to be stored in the index.
+
+    Returns
+    IntervalIndex
+    """
+
+    def __new__(
+        cls, data=None, closed=None, dtype=None, copy=False, name=None,
+    ) -> "IntervalIndex":
+        if copy:
+            data = column.as_column(data, dtype=dtype).copy(deep=True)
+        out = Frame.__new__(cls)
+        kwargs = _setdefault_name(data, name=name)
+        if isinstance(data, IntervalColumn):
+            data = data
+        elif isinstance(data, pd.Series) and (is_interval_dtype(data.dtype)):
+            data = column.as_column(pa.Array.from_pandas(data), dtype=dtype,)
+        elif isinstance(data, (pd._libs.interval.Interval, pd.IntervalIndex)):
+            data = column.as_column(data, dtype=dtype,)
+        else:
+            if data is not None and data[0].closed != closed:
+                # when closed is not the same as the data's closed
+                # we need to change the data closed value,
+                # not sure how to do this on the gpu?
+                data = pd.arrays.IntervalArray(data, closed=closed)
+            data = column.as_column(
+                data, dtype="interval" if dtype is None else dtype
+            )
+            # dtype has already been taken care
+            dtype = None
```

Comment:
Why is this necessary?

Reply:
I don't think it necessarily needs to be there. Not sure if I was getting errors because of this. I'll double-check